[JIT] Allow empty temporary list literals to be matched to arbitrary types

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74785
* __->__ #74768

As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

Differential Revision: [D35362760](https://our.internmc.facebook.com/intern/diff/D35362760)

Commit:
Allow empty temporrary list literalsto be matched to arbitrary types

[ghstack-poisoned]

Commit:
Update on "Allow empty temporrary list literalsto be matched to arbitrary types"




[ghstack-poisoned]

Commit:
Update on "[JIT] Allow empty temporary list literals to be matched to arbitrary types"


As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow empty temporary list literals to be matched to arbitrary types"


As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow empty temporary list literals to be matched to arbitrary types"


As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

[ghstack-poisoned]

Chillee commented on torch/csrc/jit/frontend/ir_emitter.cpp:
```diff
@@ -4199,6 +4203,26 @@ struct to_ir {
+    //    b.append(2)
+    // else:
+    //    a.append("hi")
+    // This is also the same behavior that C++ allows with {}
```
Probably worth mentioning here that these are intended to be cleaned up.

Chillee reviewed (approved):
Looks good! I'll approve it but might be better for somebody else familiar with Torchscript to take a look.

Commit:
Update on "[JIT] Allow empty temporary list literals to be matched to arbitrary types"


As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

[ghstack-poisoned]

Commit:
Update on "[JIT] Allow empty temporary list literals to be matched to arbitrary types"


As commented in code:
```
    // Empty List Literals that are not assigned to variables
    // may match to any list type in schema matching,
    // but still default to List[Tensor] if assigned to a variable
    // or returned from a function
    // Restricting empty list matching to temporary values
    // avoids difficult to handle cases such as
    // a = []
    // b = a
    // if cond:
    //    b.append(2)
    // else:
    //    a.append("hi")
    // This is also the same behavior that C++ allows with {}
    // (cannot assign to a variable typed as auto)
```

Fix for https://github.com/facebookresearch/torchdynamo/issues/95

Differential Revision: [D35362760](https://our.internmc.facebook.com/intern/diff/D35362760)

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35362760).
